browser/ui/webui/federated_learning_internals_ui.cc

```diff
@@ -0,0 +1,128 @@
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/grit/browser_resources.h"
+#include "chrome/grit/generated_resources.h"
+#include "components/grit/brave_components_resources.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/web_contents.h"
+#include "content/public/browser/web_ui.h"
+#include "content/public/browser/web_ui_data_source.h"
+
+FederatedLearningInternalsHandler::FederatedLearningInternalsHandler(
+    Profile* profile) {
+  data_store_service_ =
+      brave_federated::BraveFederatedServiceFactory::GetForBrowserContext(profile)
+          ->GetDataStoreService();
+}
+
+FederatedLearningInternalsHandler::~FederatedLearningInternalsHandler() =
+    default;
+
+void FederatedLearningInternalsHandler::RegisterMessageCallbacks() {
+  RegisterMessageCallback(
+      "requestAdStoreInfo",
+      base::BindRepeating(
+          &FederatedLearningInternalsHandler::OnAdStoreRequestInfo,
+          base::Unretained(this)));
+}
+
+void FederatedLearningInternalsHandler::OnAdStoreRequestInfo(
+    const base::ListValue* args) {
+  auto* const ad_notification_data_store =
+      data_store_service_->GetAdNotificationTimingDataStore();
+
+  ad_notification_data_store
+      ->AsyncCall(&brave_federated::AdNotificationTimingDataStore::LoadLogs)
+      .Then(base::BindOnce(
+          &FederatedLearningInternalsHandler::OnAdStoreLogsReceivedCallback,
+          base::Unretained(this)));
+}
+
+void FederatedLearningInternalsHandler::OnAdStoreLogsReceivedCallback(
+    brave_federated::AdNotificationTimingDataStore::
+        IdToAdNotificationTimingTaskLogMap ad_notification_timing_logs) {
+  auto list_value = std::make_unique<base::ListValue>();
+  for (auto const& object : ad_notification_timing_logs) {
+    auto dict = std::make_unique<base::DictionaryValue>();
+    auto log = object.second;
+    dict->SetIntKey("id", log.id);
+    dict->SetDoubleKey("time", log.time.ToJsTime());
+    dict->SetStringKey("locale", log.locale);
+    dict->SetIntKey("number_of_tabs", log.number_of_tabs);
+    dict->SetBoolKey("label", log.label);
+    list_value->Append(std::move(dict));
+  }
+  SendMessageToJs("adsTimingDataStoreLogsLoaded", *list_value);
+}
+
+void FederatedLearningInternalsHandler::SendMessageToJs(
+    const std::string& message,
+    const base::Value& value) {
+  const char func[] = "cr.webUIListenerCallback";
+  base::Value message_data(message);
+  std::vector<const base::Value*> args{&message_data, &value};
+  CallJavascriptFunction(func, args);
+}
+
+void FederatedLearningInternalsHandler::RegisterMessageCallback(
+    const std::string& message,
+    const MessageCallback& callback) {
+  web_ui()->RegisterDeprecatedMessageCallback(message, callback);
+}
+
+void FederatedLearningInternalsHandler::CallJavascriptFunction(
+    const std::string& function_name,
+    const std::vector<const base::Value*>& args) {
+  web_ui()->CallJavascriptFunctionUnsafe(function_name, args);
```

Comment:
this is discouraged, as the comment says.
Basically I'd recommend to use mojo approach for webui's, there are many examples across the codebase